common/appscale/common/async_retrying.py

```diff
@@ -0,0 +1,305 @@
+    if max_retries is not MISSED:
+      self.max_retries = max_retries
+    if retrying_timeout is not MISSED:
+      self.retrying_timeout = retrying_timeout
+    if retry_on_exception is not MISSED:
+      self.retry_on_exception = retry_on_exception
+
+    if func is None:
+      return self
+
+    @functools.wraps(func)
+    @gen.coroutine
+    def persistent_execute(*args, **kwargs):
+      # Allow runtime customization of retrying parameters.
+      base = kwargs.pop('backoff_base', self.backoff_base)
+      multiplier = kwargs.pop('backoff_multiplier', self.backoff_multiplier)
+      threshold = kwargs.pop('backoff_threshold', self.backoff_threshold)
+      retries_limit = kwargs.pop('max_retries', self.max_retries)
+      timeout = kwargs.pop('retrying_timeout', self.retrying_timeout)
+      check_exception = kwargs.pop('retry_on_exception',
+                                   self.retry_on_exception)
+      if check_exception is None:
+        check_exception = lambda error: True
+      elif isinstance(check_exception, (list, tuple)):
+        check_exception = lambda error: error.__class__ in check_exception
+
+      retries = 0
+      backoff = multiplier
+      start_time = time.time()
+      node_lock = self._locks[node]
+      result = None
+
+      # Wake older update calls (*)
+      node_lock.condition.notify_all()
+      node_lock.waiters += 1
+      with (yield node_lock.lock.acquire()):
+        node_lock.waiters -= 1
+
+        while True:
+          # Start of retrying iteration
+          try:
+            result = func(*args, **kwargs)
+            if isinstance(result, gen.Future):
+              result = yield result
+            break
+
+          except Exception as e:
+            retries += 1
+
+            # Check if need to retry
+            if retries_limit is not None and retries > retries_limit:
+              logging.error(
+                "Giving up retrying after {} attempts during {:0.1f}s"
+                .format(retries, time.time() - start_time))
+              fail = True
+            elif timeout and time.time() - start_time > timeout:
+              logging.error(
+                "Giving up retrying after {} attempts during {:0.1f}s"
+                .format(retries, time.time() - start_time))
+              fail = True
+            elif not check_exception(e):
+              fail = True
+            else:
+              fail = False
+
+            if fail:
+              if not node_lock.waiters:
+                del self._locks[node]
+              raise
+
+            # Proceed with exponential backoff
+            backoff = backoff * base
+            if backoff > threshold:
+              backoff = threshold
+            sleep_time = backoff * (random.random() * 0.3 + 0.85)
+            # Report problem to logs
+            stacktrace = traceback.format_exc()
+            msg = "Retry #{} in {:0.1f}s".format(retries, sleep_time)
+            logging.warn(stacktrace + msg)
```

Comment:
Use `warning` for forward compatibility since `warn` is deprecated.

Reply:
Thanks, didn't know about it